src/screen-cast.c

```diff
@@ -624,12 +929,123 @@ collect_screen_cast_stream_data (GVariantIter *streams_iter)
   return streams;
 }
 
+static void
+generate_and_save_restore_token (ScreenCastSession *screen_cast_session)
+{
+  Session *session = (Session *)screen_cast_session;
+
+  if (!screen_cast_session->restore_data)
+    {
+      if (screen_cast_session->restore_token)
+        {
+          delete_persistent_permissions (session->app_id, screen_cast_session->restore_token);
+          delete_transient_permissions (session->sender, screen_cast_session->restore_token);
+        }
+
+      g_clear_pointer (&screen_cast_session->restore_token, g_free);
+      return;
+    }
+
+  switch (screen_cast_session->persist_mode)
+    {
+    case PERSIST_MODE_NONE:
+      if (screen_cast_session->restore_token)
+        {
+          delete_persistent_permissions (session->app_id, screen_cast_session->restore_token);
+          delete_transient_permissions (session->sender, screen_cast_session->restore_token);
+        }
+
+      g_clear_pointer (&screen_cast_session->restore_token, g_free);
+      g_clear_pointer (&screen_cast_session->restore_data, g_variant_unref);
+      break;
+
+    case PERSIST_MODE_TRANSIENT:
+      if (screen_cast_session->restore_token == NULL)
+        screen_cast_session->restore_token = g_uuid_string_random ();
```

Comment:
The docs for `g_uuid_string_random()` say it has the same randomness guarantees as `GRand`, which can fallback to using the current time as a seed if `/dev/urandom` isn't available. I guess we're not worried since it's an unlikely attack vector?

Reply:
I am not completely sure this can't be abused by a malicious app, but there are a few points I had in mind when writing this code:

 * As Simon said in the first comments of this PR, there is nothing we can do against malicious apps running unsandboxed on the host. They can do literally anything to these permissions.
 * `xdg-desktop-portal` checks the app id of the app that is requesting the permission, and stores the tuple `(restore-token, app-id)`. For a malicious app to breach any security using this feature, it'd have to lie about it's app id, which either requires not being sandboxed, or tricking users with a fake app id.
 *  Even if a malicious app attempts to brute-force every possible restore token, they still wouldn't be able to predict *what* that restore token will restore (this feature guarantees that apps never know what is being shared with them; they can only restore previous sessions, but they don't know what these sessions had.)